### ISSUE TITLE
registerConnection's use of _.cloneDeep is causing ssl to fail

### ISSUE THREAD

**Issue by cludden** [balderdashy/waterline#1073]
When attempting to connect to a Google CloudSQL database using the sails-mysql adapter and a mysql with ssl connection definition:

```
module.exports.connections = {
    mysql: {
        adapter: 'sails-mysql',
        host: process.env.CLOUD_SQL_HOST,
        user: process.env.CLOUD_SQL_USER,
        database: process.env.CLOUD_SQL_DB,
        ssl: {
            key: fs.readFileSync(path.join(cloudsql, 'client-key.pem')),
            cert: fs.readFileSync(path.join(cloudsql, 'client-cert.pem')),
            ca: fs.readFileSync(path.join(cloudsql, 'server-ca.pem'))
        }
    },
```

the connection is failing with the following error:
```
error: Error (E_UNKNOWN) :: Encountered an unexpected error
: Could not connect to MySQL:
Error: unable to verify the first certificate
    at afterwards (/Users/cludden/Documents/Projects/wccsails2/node_modules/sails-mysql/lib/connections/spawn.js:72:13)
```

after much sifting, this seems to be due to the fact that the ssl file buffers returned from fs.readFileSync are being converted into js objects when cloned via the _.cloneDeep method:

adding a couple of log statements to waterline.js:
```
// lib/waterline.js

        // Call the registerConnection method
        console.log(config);
        console.log(_.cloneDeep(config));
        connection._adapter.registerConnection(_.cloneDeep(config), usedSchemas, function(err) {
          if(err) return nextItem(err);
          nextItem();
        });
```

yields the following:
```
{ pool: true,
  connectionLimit: 5,
  waitForConnections: true,
  adapter: 'sails-mysql',
  host: ********,
  user: ********,
  database: ********,
  ssl: 
   { key: <Buffer 2d 2d 2d //etc ... >,
     cert: <Buffer 2d 2d 2d //etc... >,
     ca: <Buffer 2d 2d 2d 2//etc ... > },
  identity: 'mysql' }
```

followed by:
```
{ pool: true,
  connectionLimit: 5,
  waitForConnections: true,
  adapter: 'sails-mysql',
  host: *******,
  user: *******,
  database: *******,
  ssl: 
   { key: 
      { '0': 45,
        '1': 45,
        '2': 45,
        '3': 45,
        '4': 45,
        '5': 66,
        '6': 69,
        '7': 71,
        '8': 73,
        '9': 78,
        '10': 32,
        // etc
        length: 1674,
        parent: [Object] },
     cert: 
      { '0': 45,
        '1': 45,
        '2': 45,
        '3': 45,
        '4': 45,
        '5': 66,
        '6': 69,
        '7': 71,
        '8': 73,
        '9': 78,
        '10': 32,
        // etc
        length: 1146,
        parent: [Object] },
     ca: 
      { '0': 45,
        '1': 45,
        '2': 45,
        '3': 45,
        '4': 45,
        '5': 66,
        '6': 69,
        '7': 71,
        '8': 73,
        '9': 78,
        // etc
        length: 1146,
        parent: [Object] } },
  identity: 'mysql' }
```

I am able to connect successfully using the original config definition and underlaying node-mysql module. Is there any way to avoid the use of cloneDeep here? Or prevent it from converting the file buffers to json objects?

**Comment by cludden** [comment]
I was able to fix the issue by adding the following:

/lib/waterline.js
```
// clone the config settings in a buffer-safe way
var clonedConfig = _.cloneDeep(config, function(value) {
  if (value instanceof Buffer) {
     return value.slice();
   }
});
```

right above the adapter.registerConnection call:

/lib/waterline.js 
```
// clone the config settings in a buffer-safe way
        var clonedConfig = _.cloneDeep(config, function(value) {
          if (value instanceof Buffer) {
            return value.slice();
          }
        });

        // Call the registerConnection method
        connection._adapter.registerConnection(clonedConfig, usedSchemas, function(err) {
          if(err) return nextItem(err);
          nextItem();
        });
```

Thoughts?

**Comment by devinivy** [comment]
Does it need to be cloned at all?

**Comment by cludden** [comment]
It doesn't appear like it. But I'm not super familiar with the waterline source code and if there would be issues with an adapter modifying the config options after registerConnection is called. To answer your question, passing the original config option that is built within the initialize method is working fine for me.. but maybe someone more familiar than me can chime in?

**Comment by particlebanana** [comment]
Don't you want to send the value of the files instead of the raw file buffer?

```javascript
fs.readFileSync(path.join(cloudsql, 'client-key.pem'), { encoding: 'utf8' })
```

**Comment by devinivy** [comment]
node-mysql looks like it will accept a buffer there if that's what you give it: https://github.com/felixge/node-mysql/#ssl-options

I guess the question is, should the adapters be able to modify `config` if they'd like to?  And what about the possibility of configurations that rely on non-plain objects?  It's certainly desirable to not have to clone the config if it's not necessary.

**Comment by tjwebb** [comment]
Adapters can already modify/transform the config however it wants. That cloning is happening as the config object is passed into `registerConnection`, which is where the deep cloning is causing the loss of type precision. I'm happy to remove it, I don't think it offers any benefit here.

**Comment by dmarcelino** [comment]
> we could just have an integration test that catches that sort of thing by testing for config equality before and after each adapter method call.

Sounds like a reasonable test to add to waterline-adapter-tests.

**Comment by particlebanana** [comment]
merged #1074 which should fix this.
